kani-compiler/src/kani_middle/points_to/points_to_analysis.rs

```diff
@@ -0,0 +1,626 @@
+//! necessary aliasing information for instrumenting delayed UB later on.
+//!
+//! The analysis uses Rust's dataflow framework by implementing appropriate traits to leverage the
+//! existing fixpoint solver infrastructure. The main trait responsible for the dataflow analysis
+//! behavior is `rustc_mir_dataflow::Analysis`: it provides two methods that are responsible for
+//! handling statements and terminators, which we implement.
+//!
+//! The analysis proceeds by looking at each instruction in the dataflow order and collecting all
+//! possible aliasing relations that the instruction introduces. If a terminator is a function call,
+//! the analysis recurs into the function and then joins the information retrieved from it into the
+//! original graph.
+//!
+//! For each instruction, the analysis first resolves dereference projections for each place to
+//! determine which places it could point to. This is done by finding a set of successors in the
+//! graph for each dereference projection.
+//!
+//! Then, the analysis adds the appropriate edges into the points-to graph. It proceeds until there
+//! is no new information to be discovered.
+//!
+//! Currently, the analysis is not field-sensitive: e.g., if a field of a place aliases to some
+//! other place, we treat it as if the place itself aliases to another place.
+
+use crate::kani_middle::{
+    points_to::{MemLoc, PointsToGraph},
+    reachability::CallGraph,
+    transform::RustcInternalMir,
+};
+use rustc_ast::Mutability;
+use rustc_middle::{
+    mir::{
+        BasicBlock, BinOp, Body, CallReturnPlaces, Location, NonDivergingIntrinsic, Operand, Place,
+        ProjectionElem, Rvalue, Statement, StatementKind, Terminator, TerminatorEdges,
+        TerminatorKind,
+    },
+    ty::{Instance, InstanceKind, List, ParamEnv, TyCtxt, TyKind},
+};
+use rustc_mir_dataflow::{Analysis, AnalysisDomain, Forward, JoinSemiLattice};
+use rustc_smir::rustc_internal;
+use rustc_span::source_map::Spanned;
+use std::collections::HashSet;
+
+/// Main points-to analysis object.
+struct PointsToAnalysis<'a, 'tcx> {
+    instance: Instance<'tcx>,
+    body: &'a Body<'tcx>,
+    tcx: TyCtxt<'tcx>,
+    /// This will be used in the future to resolve function pointer and vtable calls. Currently, we
+    /// can resolve call graph edges just by looking at the terminators and erroring if we can't
+    /// resolve the callee.  
+    call_graph: &'a CallGraph,
+    /// This graph should contain a subset of the points-to graph reachable from function arguments.
+    /// For the entry function it will be empty (as it supposedly does not have any parameters).
+    initial_graph: PointsToGraph<'tcx>,
+}
+
+/// Public points-to analysis entry point. Performs the analysis on a body, outputting the graph
+/// containing aliasing information of the body itself and any body reachable from it.
+pub fn run_points_to_analysis<'tcx>(
```

Comment:
can you make that interface stable please?

Reply:
The problem with this is that `PointsToGraph` uses internal types inside because `StablePlace` does not implement `Hash`, so it cannot be put inside a `HashMap`. The best thing I could do here is to restrict all use of internal types to `points_to` module through multiple stable interfaces.